File: ttd_counter/mortality_model.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal
import math


Sex = Literal["male", "female", "other"]
HealthStatus = Literal["excellent", "good", "average", "poor"]
SmokingStatus = Literal["never", "former", "current"]


@dataclass
class PersonProfile:
    age: float  # years
    sex: Sex
    height_cm: float
    weight_kg: float
    health: HealthStatus
    smoking: SmokingStatus

    @property
    def bmi(self) -> float:
        h_m = self.height_cm / 100.0
        if h_m <= 0:
            return 0.0
        return self.weight_kg / (h_m * h_m)


@dataclass
class MakehamParameters:
    """
    Parameters for a very simplified Makeham-style force of mortality:

        mu(x) = A + B * c**x

    Units are per-year hazards. These are NOT calibrated to real data and are
    intentionally rough.
    """

    A: float = 0.00027
    B: float = 0.000003
    c: float = 1.124


def base_parameters_for_sex(sex: Sex) -> MakehamParameters:
    """
    Very rough sex-based baseline.
    """
    if sex == "female":
        return MakehamParameters(A=0.00024, B=0.0000027, c=1.124)
    if sex == "male":
        pass
    return MakehamParameters()


def risk_multiplier_for_profile(profile: PersonProfile) -> float:
    """
    Crude risk multiplier based on health, BMI, and smoking.
    """
    multiplier = 1.0

    # Overall self-reported health
    if profile.health == "excellent":
        multiplier *= 0.8
    elif profile.health == "good":
        multiplier *= 0.9
    elif profile.health == "average":
        multiplier *= 1.0
    elif profile.health == "poor":
        multiplier *= 1.3

    # Smoking status
    if profile.smoking == "never":
        multiplier *= 1.0
    elif profile.smoking == "former":
        multiplier *= 1.15
    elif profile.smoking == "current":
        multiplier *= 1.6

    bmi = profile.bmi
    if bmi <= 0:
        pass
    elif bmi < 18.5:  # underweight
        multiplier *= 1.1
    elif 18.5 <= bmi <= 27:  # roughly normal / slightly overweight
        multiplier *= 0.95
    elif 27 < bmi <= 32:
        multiplier *= 1.2
    else:  # > 32 obese
        multiplier *= 1.4

    return multiplier


def makeham_mu(x: float, params: MakehamParameters) -> float:
    """
    Force of mortality at age x.
    """
    return params.A + params.B * (params.c ** x)


def expected_remaining_life_years(
    profile: PersonProfile,
    max_age: int = 110,
    step_years: float = 0.25,
) -> float:
    """
    Numerically approximate the expected remaining lifetime (in years) from
    current age using a Makeham-style hazard.

    We compute:
        E[T] ~ ∫_0^{max_age-age} S(age + t) dt
    with S approximated using small time steps and an exponential survival
    approximation S(t+dt) ~ S(t) * exp(-mu * dt).
    """
    if profile.age >= max_age:
        return 0.0

    params = base_parameters_for_sex(profile.sex)
    risk_mult = risk_multiplier_for_profile(profile)

    age = profile.age
    S = 1.0  # survival at t = 0 (conditioning on being alive now)
    t = 0.0
    e_remaining = 0.0

    while age + t < max_age and S > 1e-6:
        # Hazard at current age
        mu = makeham_mu(age + t, params) * risk_mult
        # Add contribution to expectation over this small interval
        e_remaining += S * step_years
        # Update survival over the step
        S *= math.exp(-mu * step_years)
        t += step_years

    return e_remaining

File: frontend/src/mortalityModel.ts
/**
 * Makeham-style mortality model (toy). Ported from Python for frontend-only use.
 * Not calibrated to real data; not medical or actuarial advice.
 */

export type Sex = "male" | "female" | "other";
export type HealthStatus = "excellent" | "good" | "average" | "poor";
export type SmokingStatus = "never" | "former" | "current";

export interface PersonProfile {
  age: number;
  sex: Sex;
  height_cm: number;
  weight_kg: number;
  health: HealthStatus;
  smoking: SmokingStatus;
}

interface MakehamParams {
  A: number;
  B: number;
  c: number;
}

function bmi(profile: PersonProfile): number {
  const hM = profile.height_cm / 100;
  if (hM <= 0) return 0;
  return profile.weight_kg / (hM * hM);
}

function baseParamsForSex(sex: Sex): MakehamParams {
  if (sex === "female") {
    return { A: 0.00024, B: 0.0000027, c: 1.124 };
  }
  return { A: 0.00027, B: 0.000003, c: 1.124 };
}

function riskMultiplier(profile: PersonProfile): number {
  let mult = 1.0;

  const baseHealthMult =
    profile.health === "excellent"
      ? 0.8
      : profile.health === "good"
        ? 0.9
        : profile.health === "average"
          ? 1.0
          : 1.3;
  // Progressive system: starting at age 65, increase the health effect
  // slightly every 5 years (65–69, 70–74, ...).
  let ageSteps = 0;
  if (profile.age >= 65) {
    ageSteps = Math.floor((profile.age - 65) / 5) + 1; // 65–69 => 1, 70–74 => 2, etc.
  }
  const healthEffectScale = 1 + 0.1 * ageSteps; // +10% effect per 5-year step
  const healthMult = 1 + (baseHealthMult - 1) * healthEffectScale;
  mult *= healthMult;

  if (profile.smoking === "never") mult *= 1.0;
  else if (profile.smoking === "former") mult *= 1.15;
  else if (profile.smoking === "current") mult *= 1.6;

  const b = bmi(profile);
  if (b <= 0) {
    // no change
  } else if (b < 18.5) mult *= 1.1;
  else if (b <= 27) mult *= 0.95;
  else if (b <= 32) mult *= 1.2;
  else mult *= 1.4;

  return mult;
}

function makehamMu(x: number, params: MakehamParams): number {
  return params.A + params.B * Math.pow(params.c, x);
}

/**
 * Expected remaining life years (numerical integration).
 */
export function expectedRemainingLifeYears(
  profile: PersonProfile,
  maxAge: number = 110,
  stepYears: number = 0.25
): number {
  if (profile.age >= maxAge) return 0;

  const params = baseParamsForSex(profile.sex);
  const riskMult = riskMultiplier(profile);

  let age = profile.age;
  let S = 1.0;
  let t = 0;
  let eRemaining = 0;

  while (age + t < maxAge && S > 1e-6) {
    const mu = makehamMu(age + t, params) * riskMult;
    eRemaining += S * stepYears;
    S *= Math.exp(-mu * stepYears);
    t += stepYears;
  }

  return eRemaining;
}

/**
 * Estimated date of death from remaining years (365.25 days/year).
 */
export function estimateDeathDate(remainingYears: number): string {
  const now = new Date();
  const days = Math.floor(remainingYears * 365.25);
  const d = new Date(now.getTime() + days * 24 * 60 * 60 * 1000);
  return d.toISOString().slice(0, 10);
}

const MAX_AGE = 110;
const STEP_YEARS = 0.25;

/**
 * Survival curve S(t) = P(alive at current age + t | alive at current age).
 * Returns points for plotting: t (years from now), S (probability).
 */
export function getSurvivalCurve(
  profile: PersonProfile,
  maxYears: number = 60
): { t: number; S: number }[] {
  if (profile.age >= MAX_AGE) return [{ t: 0, S: 1 }, { t: maxYears, S: 0 }];

  const params = baseParamsForSex(profile.sex);
  const riskMult = riskMultiplier(profile);

  const points: { t: number; S: number }[] = [{ t: 0, S: 1 }];
  let age = profile.age;
  let S = 1.0;
  let t = STEP_YEARS;

  while (t <= maxYears && age + t < MAX_AGE && S > 1e-6) {
    const mu = makehamMu(age + t, params) * riskMult;
    S *= Math.exp(-mu * STEP_YEARS);
    points.push({ t, S });
    t += STEP_YEARS;
  }

  return points;
}
